dynamiq/nodes/tools/mcp_adapter.py

```diff
@@ -0,0 +1,248 @@
+            raise ToolExecutionException(
+                f"Tool '{self.name}' failed to call tool from the MCP server."
+                f"Error: {str(e)}. Please analyze the error and take appropriate action.",
+                recoverable=True,
+            )
+
+        logger.info(f"Tool {self.name} - {self.id}: finished with result:\n{str(result)[:200]}...")
+        return {"content": result}
+
+
+class MCPAdapterTool(Node):
+    """
+    A tool that manages connections to MCP servers and initializes MCP tools.
+
+    Attributes:
+      group (Literal[NodeGroup.TOOLS]): Node group.
+      name (str): Node name.
+      description (str): Node description.
+      mcp_tools (list): A list of initialized MCP tools.
+      server_params (ServerParameters): The parameters for connecting to the MCP server.
+    """
+
+    group: Literal[NodeGroup.TOOLS] = NodeGroup.TOOLS
+    name: str = "MCP Adapter Tool"
+    description: str = "The tool used to initialize available MCP tools based on provided server parameters."
+
+    mcp_tools: list[MCPTool] = field(default_factory=list)
+    server_params: ServerParameters
+
+    async def initialize_tools(self):
+        """
+        Initializes the MCP tool list from the client session.
+
+        Returns:
+            None
+        """
+        async with await get_client(self.server_params) as (read, write):
+            async with ClientSession(read, write) as session:
+                await session.initialize()
+                tools = await session.list_tools()
+                self.mcp_tools = [
+                    MCPTool(
+                        name=tool.name,
+                        description=tool.description,
+                        input_schema=tool.inputSchema,
+                        server_params=self.server_params,
+                    )
+                    for tool in tools.tools
+                ]
+
+        logger.info(f"Tool {self.name}: {len(self.mcp_tools)} MCP tools initialized from a server.")
+
+    def get_mcp_tools(self) -> list[MCPTool]:
+        """
+        Synchronously fetches and initializes MCP tools if not already available.
+
+        Returns:
+            list[MCPTool]: A list of initialized MCPTool instances.
+        """
+        if is_called_from_async_context():
+            with ThreadPoolExecutor() as executor:
+                future = executor.submit(lambda: asyncio.run(self.get_mcp_tools_async()))
+                return future.result()
+        return asyncio.run(self.get_mcp_tools_async())
+
+    async def get_mcp_tools_async(self) -> list[MCPTool]:
+        """
+        Asynchronously fetches and initializes MCP tools if not already available.
+
+        Returns:
+            list[MCPTool]: A list of initialized MCPTool instances.
+        """
+        if not self.mcp_tools:
+            await self.initialize_tools()
+        return self.mcp_tools
+
+    def execute(self, **kwargs):
```

Comment:
Can we also here return list of initialized MCP tools? Maybe under some dict key